pkg/engine/executor/parse_logfmt.go

```diff
@@ -0,0 +1,189 @@
+func parseRequestedKeys(input *array.String, requestedKeys []string, columnBuilders map[string]*array.StringBuilder, allocator memory.Allocator) ([]string, []error, bool) {
+	columnOrder := make([]string, 0, len(requestedKeys))
+	errorList := make([]error, 0, input.Len())
+	hasAnyError := false
+
+	// Pre-create all builders
+	for _, key := range requestedKeys {
+		columnBuilders[key] = array.NewStringBuilder(allocator)
+		columnOrder = append(columnOrder, key)
+	}
+
+	// Process each line
+	for i := 0; i < input.Len(); i++ {
+		line := input.Value(i)
+		parsed, err := TokenizeLogfmt(line, requestedKeys) // Already filtered
+		errorList = append(errorList, err)
+		if err != nil {
+			hasAnyError = true
+		}
+
+		// Append values for all requested keys
+		for _, key := range requestedKeys {
+			if value, ok := parsed[key]; ok {
+				columnBuilders[key].Append(value)
+			} else {
+				columnBuilders[key].AppendNull()
+			}
+		}
+	}
+
+	return columnOrder, errorList, hasAnyError
+}
+
+// parseAllKeys handles the dynamic path when no specific keys are requested
+// Discovers columns dynamically as lines are parsed
+func parseAllKeys(input *array.String, columnBuilders map[string]*array.StringBuilder, allocator memory.Allocator) ([]string, []error, bool) {
+	columnOrder := []string{}
+	errorList := make([]error, 0, input.Len())
+	hasAnyError := false
+
+	for i := 0; i < input.Len(); i++ {
+		line := input.Value(i)
+		parsed, err := TokenizeLogfmt(line, nil) // Get all keys
+		errorList = append(errorList, err)
+		if err != nil {
+			hasAnyError = true
+		}
+
+		// Track which keys we've seen this row
+		seenKeys := make(map[string]bool)
+
+		// Add values for parsed keys
+		for key, value := range parsed {
+			seenKeys[key] = true
+			builder, exists := columnBuilders[key]
+			if !exists {
+				// New column discovered - create and backfill
+				builder = array.NewStringBuilder(allocator)
+				columnBuilders[key] = builder
+				columnOrder = append(columnOrder, key)
+
+				// Backfill NULLs for previous rows
+				for j := 0; j < i; j++ {
+					builder.AppendNull()
+				}
+			}
+			builder.Append(value)
+		}
+
+		// Append NULLs for columns not in this row
+		for _, key := range columnOrder {
+			if !seenKeys[key] {
+				columnBuilders[key].AppendNull()
+			}
+		}
+	}
+
+	// Sort column order for consistency
+	sort.Strings(columnOrder)
+
+	return columnOrder, errorList, hasAnyError
+}
+
+// TokenizeLogfmt parses logfmt input using the standard decoder
+// Returns a map of key-value pairs with last-wins semantics for duplicates
+// If requestedKeys is provided, the result will be filtered to only include those keys
+func TokenizeLogfmt(input string, requestedKeys []string) (map[string]string, error) {
+	result := make(map[string]string)
+	decoder := logfmt.NewDecoder(unsafeBytes(input))
+
+	// Parse all key-value pairs
+	for !decoder.EOL() && decoder.ScanKeyval() {
+		key := string(decoder.Key())
+		value := string(decoder.Value())
```

Comment:
These will create a copy of the byte slice, you may want to use an `unsafeString` here to avoid the copy similar to what you do in creating the decoder.